### PR TITLE
feat(plugin-page-builder): let a plugin contribute blocks through the registry

### PR DESCRIPTION
**Pull request by mobeenabdullah**
Plan-01's PR-7, re-scoped from `feat(nextly)` to `feat(plugin-page-builder)`. Full audit and decision record: `tasks/2026-07-31-pr7-contributes-blocks-audit.md`; plan: `tasks/2026-07-31-pr7-implementation-plan.md`.

## What a contributing plugin writes

```ts
definePlugin({
  name: "@acme/pricing-blocks",
  init: ctx => blockRegistry(ctx).register([pricingTable, pricingTier]),
})
```

`defineBlock` and the block types come from `@nextlyhq/plugin-page-builder`, so a block author takes one package.

**Core is unchanged.** No `contributes.blocks` key, no dependency on the block engine.

## Why the plan's original scope was wrong

PR-7 was written 2026-07-24 and says `contributes.blocks` lives in core. The decision that moved blocks out of core was taken 2026-07-28, and #425 removed core's engine dependency entirely. Implementing PR-7 as written would have re-added it.

The packaging decision PB-D10 had already settled this on 2026-07-22 — "only `plugin-page-builder` is a Nextly plugin, and it depends on the other three" — so core was never meant to be in the blocks dependency graph. The scope line was stale, not the feature.

## Why a service rather than a direct import

The engine's registry is cleared and rebuilt on every boot. A contributor importing `registerBlocks` directly can register before the rebuild and lose its blocks **with no error** — the same class of bug the field-type registry avoided by having a defined drain point.

Resolving a service cannot fail that way, and the boot sequence guarantees it (`di/register.ts:2383`):

> PASS 1 — ... Services register BEFORE any init runs, so a plugin's `init` can resolve any other plugin's service lazily via `ctx.services.plugins.<name>.<svc>`, regardless of init order.

The first test asserts exactly this by listing the contributor **before** the page builder.

## DX: a typed accessor, not a cast

`ctx.services.plugins` is typed `Record<string, Record<string, unknown>>`, so reaching it directly costs a cast at every call site. `blockRegistry(ctx)` resolves it once and states what came back, so authors get completion and argument checking — and a named error when the page builder is not installed, rather than a generic missing-service message.

## Stability

`contributes.services` was recorded in `STABILITY.md` as "no first-party plugin contributes a service yet". That is now false, and per the ladder a surface graduates once a first-party plugin exercises it. The ledger entry is updated deliberately, including the note that a contributing plugin should export a typed accessor for its own service as this one does.

## Tests (plugin integration suite, all three dialects)

1. The block registers with the contributor listed FIRST — the ordering that breaks if registration depended on init order.
2. `getBlockSource` names the contributing plugin, so a collision between two plugins is diagnosable.
3. The page builder absent → a clear error, not a silent no-op. This is the failure the design exists to prevent, so it is asserted.
4. An empty contribution is accepted without registering anything.

Stub-verified: removing the contributed service fails 3 of the 4.

## Verification

- build 17/17 · typecheck 0 · lint 0
- plugin integration **16 passed / 6 skipped**, plugin unit **305 passed**
- `nextly` **400 failed / 37 failed files** — the known baseline, unchanged

## Not in this PR

The admin editor-component seam (PR-8), manifest generation (PR-9), and the block API freeze (PR-12, after the DX gate in PR-11).

---

## Scope correction (added after review)

Review found that this PR delivers less than the title suggests, and that is accurate.

**What it does:** a plugin can register blocks through the page builder's service, the contribution is attributed to the contributing plugin, and the registry is reset on every boot so a config reload does not fail with a collision.

**What it does NOT do:** a contributed block does not appear in the editor, and it does not gate a write. The editor and renderer read the PoC's `defaultBlockRegistry` (`core/registry.ts`, keyed on `def.type`), not the engine registry (keyed on `def.name`). Bridging them would wire the new contribution path into the legacy code the rebuild is replacing, which PB-D1 rules out.

Two follow-ups make contributions observable, and neither belongs in this PR:

- **Decision D4 — status-aware validation.** `validateBlocksValue` hardcodes `mode: "forgiving"`, where an unregistered type is a warning rather than an error. Enforcement needs effective status threaded through the write path so a publish runs strict. This was PR-6b scope.
- **PR-8 — the admin editor-component seam**, which is what puts a contributed block in front of an editor-user.

An attempt to close the first of these inside this PR was reverted: supplying `registryLookup()` to the validator changed no behaviour, because the warnings it produced were filtered out one line later. Stubbing it out left every test passing, which is how it was caught.

## Summary by CodeRabbit

* **New Features**
  * Added public blocks APIs for defining block types, rendering contracts, supports, and property schemas.
  * Added page-builder block registration across plugins, including contributor tracking and support registration.
  * Added new `./blocks` package entry points for the plugin SDK and page builder.
  * Block registrations now reset cleanly during initialization and support repeated registration without collisions.

* **Documentation**
  * Updated experimental API guidance for custom services and block functionality.

### CODE REVIEW

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/plugin-page-builder/src/blocks/registration-service.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+/** Build the service the page builder contributes. */
+export function createBlockRegistrationService(): BlockRegistrationService {
+  return {
+    register(definitions, source) {
+      const list = Array.isArray(definitions) ? definitions : [definitions];
+      // Nothing to do rather than an error: a plugin whose block list is
+      // configuration-dependent may legitimately contribute none.
+      if (list.length === 0) return;
+      registerBlocks(list, { source });
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/blocks/registration-service.ts:131]: Correct, fixed in `b84c9b172`.

Verified exactly as described: `clearBlocks` appeared only in tests, the boot path never called it, and the store is `globalThis`-pinned so it survives a reload while every plugin `init` runs again.

Cleared inside the service FACTORY rather than in the page builder `init`. Init order is not fixed, so a contributor whose init ran first would have its blocks wiped by a later clear; the factory is memoized per boot (`plugin-services-registry.ts:61` short-circuits on `entry.resolved`), so the reset happens exactly once and always before the first registration that goes through it.

You were also right that the `afterEach` hid this. The new test boots twice with no cleanup in between, and stub-verifying it reproduces your prediction precisely: `NEXTLY_BLOCK_COLLISION: block "acme/pricing-table" is already registered ... and cannot be redefined`.

[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/blocks/registration-service.ts:131]: Confirmed, and my first attempt at fixing it was wrong — thank you for the precision here, it saved shipping something worse.

Verified your claim: `EditorSurface` and the panels read `defaultBlockRegistry` from `core/registry.ts`, and nothing in the plugin runtime reads the engine registry — only types are imported from it. My test asserted `getBlock()` against the registry nothing consumes, so it proved registration happened and nothing more.

**What I tried and reverted.** I wired `registryLookup()` into `validateBlocksValue`, expecting that to be the "activation" PR-6a decision D3 described. Stubbing it out left every test passing, because `validateBlocksValue` hardcodes `mode: "forgiving"` and an unregistered type is only `severity: "error"` in `strict` (`validation.ts:175`) — so the lookup produced warnings the validator then filtered away. Inert code, so it is out; I would rather ship an honest gap than a line that looks like a fix.

**What the real fix needs.** Enforcement depends on threading effective status into the validator so a publish runs strict while a draft stays forgiving — PR-6a decision D4, which was PR-6b scope and never implemented. That is genuine work across the write path, not a line in this PR.

The two registries are also not bridgeable as they stand: `core/registry.ts` keys on `def.type` with the PoC `BlockDefinition`, the engine keys on `def.name` with `AnyBlockDefinition`. Registering contributions into the PoC registry would wire the new contribution path into the legacy code the rebuild is replacing (PB-D1: no compat layer), so that is the wrong direction.

Leaving this open deliberately. This PR delivers the contribution channel, provenance, and boot safety; it does NOT make a contributed block render or gate a write, and the PR description should say so rather than imply otherwise. Filing D4 enforcement plus the PR-8 admin seam as the work that makes contributions observable.

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/plugin-page-builder/src/index.ts [hunk 2]
```diff
@@ -54,3 +54,21 @@ export type {
...
+ * through core's public surface would put the engine's types back into an API
+ * that deliberately no longer references them.
+ */
+export {
+  blockRegistry,
+  BLOCK_SERVICE,
+  PAGE_BUILDER_PLUGIN,
+} from "./blocks/registration-service";
+export type { BlockRegistrationService } from "./blocks/registration-service";
+export { defineBlock } from "@nextlyhq/blocks-engine";
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/index.ts]: Correct, and a silent break I introduced. Fixed in `9710318b4`.

Verified: the root already does `export * from "./core"`, and an explicit re-export shadows a star export — so every consumer importing `defineBlock` from the package root would have received the engine helper, which only returns its argument, and their blocks would have stopped registering with no error at all.

`BlockDefinition` collides the same way (`core/types.ts:209` declares its own), so that was shadowing too.

Both re-exports are gone. Block authors take `defineBlock` from `@nextlyhq/blocks-engine` directly until the PoC registry is removed and the name is actually free; only `AnyBlockDefinition`, which does not collide, is re-exported. The reasoning is recorded at the export site so nobody re-adds it.

File: packages/plugin-page-builder/src/plugin.ts [hunk 3]
```diff
@@ -36,6 +40,13 @@ export const pageBuilder = (opts: PageBuilderOptions = {}) =>
         "Build pages visually from blocks with drag-and-drop editing",
     },
     contributes: {
+      // The channel another plugin adds blocks through. Core carries no
+      // `contributes.blocks` key — a plugin contributing blocks is contributing
+      // to the page builder, not to Nextly — so the registry is offered here and
+      // reached via `ctx.services.plugins`.
+      services: {
+        [BLOCK_SERVICE]: () => createBlockRegistrationService(),
+      },
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/plugin.ts:68]: Correct, fixed in `9710318b4` — and the fix is not the bump.

The declared version had fallen eight releases behind (`0.0.2-alpha.29` against a package at `0.0.2-alpha.49`), so a contributor requiring the version that actually carries this service would be refused at boot by `validatePluginVersions`.

Bumping the literal would just drift again at the next release, so I checked the siblings: `plugin-form-builder` and `plugin-seo` both read the version from `package.json`, with a comment saying "so it can never drift from the published package". The page builder was the outlier. It now reads it the same way, which removes the class of bug rather than this instance.

File: packages/plugin-page-builder/src/blocks/registration-service.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+ * every plugin's `init` runs again on that reload — so without a reset the
+ * second boot re-registers definitions that are already there and the engine
+ * refuses them as collisions. Cleared here rather than in the page builder's own
+ * `init` because init order is not fixed: a contributor whose init ran first
+ * would have its blocks wiped by a later clear. This factory is memoized per
+ * boot, so the reset happens exactly once and always before the first
+ * registration that goes through it.
+ */
+export function createBlockRegistrationService(): BlockRegistrationService {
+  clearBlocks();
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/blocks/registration-service.ts:123]: Correct, and it is the case that mattered most. Fixed in `9710318b4`.

You are right that my earlier fix only covered reloads where a contributor still exists. With the last contributor removed, nothing resolves the service, the lazy factory never runs, and precisely the definitions that should disappear are the ones that survive.

The reset moved out of the factory into the page builder `setup`, which runs at `di/register.ts:393` — before any plugin `init`, and unconditionally whether or not anything contributes.

New test boots with a contributor, then boots again without it, and asserts the block is gone. Stub-verified: removing the `setup` hook fails that test and the second-boot test, and nothing else.

**Comment by mobeenabdullah** [comment]
@codex please review this PR

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/plugin-page-builder/src/plugin.ts [hunk 5]
```diff
@@ -1,9 +1,25 @@
+import { createRequire } from "node:module";
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/plugin.ts]: Correct, and confirmed in the build output rather than by reasoning: `dist/index.js:105` carried `import { createRequire } from "module"`, reachable from the package root, which browser bundlers cannot resolve.

This was collateral from the previous fix in this same PR — I copied `plugin-form-builder`s runtime lookup without checking that its comment ("this module is not part of the admin bundle") does not hold for the page builder, whose root IS the isomorphic entry.

Replaced with a static `import { version } from "../package.json"`. The bundler inlines it, so `dist/index.js` now contains the literal `"0.0.2-alpha.49"` and zero occurrences of `createRequire` or `from "module"`. It still cannot drift from the published package, which was the point of the original change.

File: .changeset/contributes-blocks.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+"@nextlyhq/prettier-config": patch
+"@nextlyhq/telemetry": patch
+"@nextlyhq/tsconfig": patch
+---
+
+A plugin can now add its own blocks to the page builder.
+
+The page builder exposes its block registry as a service, and a contributing plugin reaches it from `init` with `blockRegistry(ctx).register(myBlocks, pluginName)`. Registering this way rather than by importing the engine is what makes the timing safe: the block registry is cleared and rebuilt on every boot, so a direct call can land before the rebuild and lose the blocks with no error, while services are recorded before any plugin's `init` runs.
+
+`defineBlock` and the block types are re-exported from `@nextlyhq/plugin-page-builder`, so a block author takes one package. Nextly core is unchanged: it carries no blocks contribution key and does not depend on the block engine, because contributing blocks is contributing to the page builder rather than to the framework.
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply .changeset/contributes-blocks.md]: Correct, fixed in `c95da1400`. The changeset described a re-export that was removed later in the same PR once it turned out to shadow the existing `defineBlock`. It now states where `defineBlock` actually comes from and why the page builder does not re-export it.

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/plugin-page-builder/src/plugin.ts [hunk 7]
```diff
@@ -16,7 +32,15 @@ export interface PageBuilderOptions {
 export const pageBuilder = (opts: PageBuilderOptions = {}) =>
   definePlugin({
     name: "@nextlyhq/plugin-page-builder",
-    version: "0.0.2-alpha.29",
+    // Runs before any plugin's `init`, and runs whether or not anything
+    // contributes — so the registry starts every boot empty even when the last
+    // contributing plugin has just been removed. The config is returned
+    // untouched; this hook is used only for the reset.
+    setup: config => {
+      resetBlockRegistry();
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/plugin.ts]: Correct, and it is a regression the previous fix in this PR introduced. Reverted in `0e3aa05f8`.

Verified both premises: `reload-config.ts:1009` states the reload "never goes back through `registerServices`", and `config-loader.ts` does re-run every plugin `setup`. So clearing from `setup` empties the registry on the first unrelated save with no `init` left to repopulate it, and contributed blocks stay gone until a restart. Strictly worse than what it replaced.

The reset is back in the service factory, and the placement is now load-bearing in both directions — not in `init`, because init order is not fixed and a contributor that ran first would be wiped; not in `setup`, for the reason above. Both constraints are recorded at the code.

**This supersedes the earlier "reset even when no contributor resolves the service" thread.** The two cannot both be satisfied: catching the removed-contributor case requires clearing somewhere `init` does not follow, which is exactly what breaks HMR. I took the smaller cost — a removed contributor leaves stale definitions until the process restarts, and since the registry lives on `globalThis` it never outlives the process. That tradeoff is documented at the reset rather than left implicit.

I also removed a test I had written for the removed-contributor case. Rewritten for the new behaviour it asserted `expect(true).toBe(true)`, which measures nothing — the tradeoff belongs in a comment, not a passing test that checks no code.

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/plugin-page-builder/src/blocks/registration-service.ts [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+ * This factory is memoized per boot, so the reset happens exactly once, in the
+ * same lifecycle as every registration that follows it.
+ *
+ * The cost is that a contributor removed mid-process leaves its blocks behind:
+ * with nothing resolving this service, nothing clears them. That is the smaller
+ * problem — the registry lives on `globalThis`, so it does not outlive the
+ * process and a restart begins empty.
+ */
+export function createBlockRegistrationService(): BlockRegistrationService {
+  clearBlocks();
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/blocks/registration-service.ts:123]: Correct, fixed in `fc4a0d86d`.

Verified: `clearBlocks` at `registry.ts:310` empties the block store AND sets `__nextly_blockSupports` to undefined, so a plugin registering a support immediately before the blocks that use it lost that support between the two calls — and any support an earlier plugin had registered went with it.

Added `clearBlockDefinitions()` to the engine, which drops only the block entries, and the service resets with that. Supports are a vocabulary blocks are validated against rather than part of the per-boot block set, so they are not the reset path own concern.

Test registers a support and a block in the same `init` and asserts both survive. Stub-verified: putting `clearBlocks` back fails that test and only that test.

File: .changeset/contributes-blocks.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+"@nextlyhq/prettier-config": patch
+"@nextlyhq/telemetry": patch
+"@nextlyhq/tsconfig": patch
+---
+
+A plugin can now add its own blocks to the page builder.
+
+The page builder exposes its block registry as a service, and a contributing plugin reaches it from `init` with `blockRegistry(ctx).register(myBlocks, pluginName)`. Registering this way rather than by importing the engine is what makes the timing safe: the block registry is cleared and rebuilt on every boot, so a direct call can land before the rebuild and lose the blocks with no error, while services are recorded before any plugin's `init` runs.
+
+`defineBlock` comes from `@nextlyhq/blocks-engine`; the page builder does not re-export it, because its package root already exports a different `defineBlock` from the PoC registry and shadowing that would silently stop existing consumers registering their blocks. Nextly core is unchanged: it carries no blocks contribution key and does not depend on the block engine, because contributing blocks is contributing to the page builder rather than to the framework.
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply .changeset/contributes-blocks.md]: Fair, and fixed in `fc4a0d86d` — though not by routing through `@nextlyhq/plugin-sdk`.

You are right that telling authors to import from the engine was wrong: it is an internal library with no stability promise, so an engine refactor could break published plugins. That instruction was an accident of the previous fix, not a decision — the intended surface was the page builder, and I dropped that re-export when it turned out to shadow the root `defineBlock`.

Now published as a SUBPATH: `@nextlyhq/plugin-page-builder/blocks` exports `defineBlock`, the block types, and `blockRegistry`. A subpath sidesteps the collision entirely, so the name stays clean and no existing consumer changes.

On the SDK specifically — that was considered and decided against by the founder (recorded in `tasks/2026-07-31-pr7-contributes-blocks-audit.md`, D2). A block is a page-builder concept, and re-exporting it from the core SDK would put the engine types back into an API that #425 deliberately stopped referencing. The page builder is a published package that states its own version, so a contributor pins compatibility with `dependsOn` exactly as for any other plugin it builds on — which is the same guarantee, sourced from the package that actually owns the concept.

The tests now import through that subpath rather than reaching inside, so the published path is the one under test.

File: packages/plugin-page-builder/src/blocks/__tests__/block-contribution.integration.test.ts [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+import { createTestNextly, type TestNextly } from "nextly/testing";
+
+import { blockRegistry } from "../registration-service";
+import { pageBuilder } from "../../plugin";
+
+let current: TestNextly | undefined;
+
+afterEach(async () => {
+  await current?.cleanup?.();
+  current = undefined;
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/blocks/__tests__/block-contribution.integration.test.ts:34]: Correct, fixed in `fc51c94e6`.

Verified: `TestNextly` exposes `destroy()` (`test-nextly.ts:317`), there is no `cleanup`, so `current?.cleanup?.()` was optional-chaining into nothing — every test left its adapter and plugin lifecycle live.

Both call sites now use `destroy()`.

Worth noting for the second-boot test, since a real teardown could have changed what it proves: `destroy()` resets singletons but does NOT clear the engine block registry, so that test still passes only because of the reset in the service factory. Re-stub-verified after the change — removing the reset still fails it, and only it.

**Comment by mobeenabdullah** [comment]
@codex please review this PR

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/plugin-page-builder/src/blocks/index.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+ * page-builder concept: routing it through the core SDK would put the engine's
+ * types back into an API that deliberately no longer references them. The page
+ * builder is a published package that states its own version, so a contributor
+ * pins compatibility with `dependsOn` exactly as it would for any other plugin
+ * it builds on.
+ *
+ * @module blocks
+ */
+
+export { defineBlock } from "@nextlyhq/blocks-engine";
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/blocks/index.ts]: You were right to push back, and I was wrong to decline this once. Fixed in `694ae25b6`.

`defineBlock` and the block definition types now come from **`@nextlyhq/plugin-sdk/blocks`**. AGENTS.md is unambiguous that the SDK is the only stable surface a plugin author is offered, and my earlier reply used decision D2 to justify a boundary D2 was not actually about — D2 chose the page builder over the CORE entry, not over the SDK.

Published as a subpath rather than on the SDK main entry, which resolves the objection I had raised: a plugin that has nothing to do with blocks never pulls the engine into its type graph, so the SDK does not grow a block vocabulary everyone carries. `@nextlyhq/blocks-engine` is an optional peer, so it is only required by packages that actually import the subpath.

What stays on `@nextlyhq/plugin-page-builder/blocks` is only the registry itself — `blockRegistry`, the plugin name, the service type — because that is this plugins own registry rather than a core contract, and it is the package a contributor already names in `dependsOn`. Contracts from the SDK, registration from the plugin that owns it.

The tests import through both published paths, so the surfaces under test are the ones a contributor uses.

File: packages/plugin-page-builder/src/blocks/registration-service.ts [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+ * This factory is memoized per boot, so the reset happens exactly once, in the
+ * same lifecycle as every registration that follows it.
+ *
+ * The cost is that a contributor removed mid-process leaves its blocks behind:
+ * with nothing resolving this service, nothing clears them. That is the smaller
+ * problem — the registry lives on `globalThis`, so it does not outlive the
+ * process and a restart begins empty.
+ */
+export function createBlockRegistrationService(): BlockRegistrationService {
+  clearBlockDefinitions();
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/blocks/registration-service.ts]: Correct — and this was caused by the fix to the previous supports finding, which traded a same-init wipe for a cross-boot collision. Fixed in `694ae25b6`.

Taking the second option from your earlier comment: support registration now goes through the same service as blocks (`registry.registerSupport(...)`), so both are contributed after the reset and both share the per-boot lifecycle. The reset is back to `clearBlocks()`, which clears both — safe now that neither can be registered before it runs.

`clearBlockDefinitions`, added for the previous fix, is removed rather than left as unused API.

You were also right that the second-boot test proved nothing about supports, since its contributor declared none. There is now a second-boot test that DOES register a support. Stub-verified by bypassing the service and calling the engine`s `registerSupport` directly, which reproduces your prediction exactly: `NEXTLY_SUPPORT_COLLISION`.

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/plugin-page-builder/src/blocks/registration-service.ts [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+ * refused as using an unknown one. Both are contributed through this service,
+ * so both are registered after the reset and neither can be caught by it.
+ *
+ * The cost is that a contributor removed mid-process leaves its blocks behind:
+ * with nothing resolving this service, nothing clears them. That is the smaller
+ * problem — the registry lives on `globalThis`, so it does not outlive the
+ * process and a restart begins empty.
+ */
+export function createBlockRegistrationService(): BlockRegistrationService {
+  clearBlocks();
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/blocks/registration-service.ts:123]: Confirmed, and fixed in 4d81e988a — you are right that the previous tree documented the leak instead of closing it.

Verified the mechanism before changing anything: the service is lazy, the reset lives in its factory, so on a boot where no plugin contributes nothing resolves it and `clearBlocks()` never runs. The departed plugin's blocks and supports stay on `globalThis` for the life of the process.

**The fix.** The page builder now resolves its own registry from `init`, so the reset is unconditional per boot rather than conditional on somebody contributing. This works because of two properties that already hold: `registerServices` calls `clearPluginServices()` at the top of every boot (`di/register.ts`), so the memoized instance is dropped and the factory re-runs; and resolution is memoized *within* a boot, so the page builder resolving from `init` cannot wipe a contributor whose `init` ran first — the factory clears only on its first resolution, whoever triggers it. Both orderings are therefore correct, which the existing order-independence test still covers.

**Stub-verified**, because a test that passes when you break its subject measures nothing. Deleting the new `init` fails exactly one test — `drops a removed contributor's blocks on the next boot` — and no other (1 failed / 14 passed, against a 15-passed baseline). Both boots run inside a single test on purpose: the suite's `afterEach` clears the registry, which would have hidden precisely this leak.

**What I did not do**, and why the wording of the remaining gap changed rather than disappeared. Your finding also names the config-reload path. A reload re-runs neither `registerServices` nor any `init` — `reload-config.ts` says so in its own comment ("never goes back through `registerServices`"). So on reload nothing re-registers contributions either, and the registry legitimately holds exactly the blocks the running process registered; a plugin removed by that reload stays visible until restart. Clearing there is what the earlier `setup` attempt did, and it broke HMR: it emptied the registry with nothing left to repopulate it. Closing that half needs core to re-run `init` on reload, which is a core change and not this PR. The comment now states that boundary precisely instead of describing the whole thing as an accepted cost.

File: packages/plugin-sdk/package.json [hunk 14]
```diff
@@ -62,6 +68,9 @@
     "@nextlyhq/admin": {
       "optional": true
     },
+    "@nextlyhq/blocks-engine": {
+      "optional": true
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-sdk/package.json]: Confirmed and fixed in bdd00a356. `@nextlyhq/blocks-engine` is now a real `dependency` of the SDK rather than an optional peer.

Verified the claim against the emitted artifact rather than the source, because that is where it bites. `packages/plugin-sdk/dist/blocks.mjs` is exactly one line:

```js
export { defineBlock } from '@nextlyhq/blocks-engine';
```

A bare runtime re-export. An optional peer means the install succeeds silently and the failure only appears when the author imports the documented path, and under a strict `node_modules` layout the page builder's own copy is nested under it and unreachable from the author's package. Telling a plugin author to install the engine directly also contradicts AGENTS.md, which makes the SDK the only stable surface they are offered.

**Corroborating evidence I did not expect to find:** `pnpm-lock.yaml` already listed the engine under `dependencies` for `packages/plugin-sdk` while `package.json` declared it an optional peer. That is why nothing failed locally or in CI — pnpm had linked it into the SDK's own `node_modules`, so every in-repo consumer resolved it and only a published external consumer would have hit the break. `pnpm install --frozen-lockfile` is a no-op after the change, so the lockfile needed no edit.

**Why a dependency and not the alternatives.** Bundling the subpath would give the SDK its own copy of `defineBlock`; it is an identity function today, so it would work, but it silently forks if the engine ever makes it registry-touching — a landmine for a one-line saving. A required (non-optional) peer would make the author install a package the docs call internal, which is the contradiction the finding raises. The engine is a dependency-free leaf, in the same lockstep fixed group, and its registry is pinned to `globalThis`, so even a duplicated install reads and writes the one registry the page builder consumes. The subpath still does its original job: block types stay off the default import, so a form-builder author never pulls the engine into their type graph.

File: packages/plugin-page-builder/src/blocks/registration-service.ts [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+  return {
+    registerSupport,
+    register(definitions, source) {
+      const list = Array.isArray(definitions) ? definitions : [definitions];
+      // Nothing to do rather than an error: a plugin whose block list is
+      // configuration-dependent may legitimately contribute none.
+      if (list.length === 0) return;
+      registerBlocks(list, { source });
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/blocks/registration-service.ts:131]: Agreed and fixed in 4d81e988a. `register()` no longer takes a `source`; `blockRegistry(ctx)` binds it from `ctx.self.name`.

Verified `ctx.self.name` is authoritative before relying on it: `resolvePluginSelf` (`packages/nextly/src/plugins/self.ts`) returns `{ name: plugin.name, ... }` and the context is built per plugin in boot pass 1, so it is the registering plugin's own resolved identity and not something a call site can restate.

Shape of the change: the service instance is shared by every contributor, so it keeps an internal two-argument form, and the per-caller handle `blockRegistry` returns is what closes over that caller's name. The public `BlockRegistrationService` is now `register(definitions)` — one argument, and no way to name the wrong plugin.

**Stub-verified, and deliberately with a stub that the pre-existing test could not catch.** Replacing `ctx.self.name` with the literal `"@acme/pricing-blocks"` leaves the old single-contributor test `records which plugin contributed the block` passing, because that is exactly the name it expects. Under that stub exactly one test fails: the new `attributes each block to the plugin that registered it`, which boots two contributors in one boot and asserts each block against its own registrar. So the new test demonstrably covers something the existing one cannot, rather than restating it.

Also updated the changeset, which still showed the two-argument call in the user-facing example.

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/plugin-sdk/src/blocks.ts [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+ * @packageDocumentation
+ */
+
+/**
+ * @experimental The block API is frozen at the end of the engine phase, not
+ *   now. Until then a contributed block may need changes when the definition
+ *   shape settles.
+ */
+export { defineBlock } from "@nextlyhq/blocks-engine";
+export type {
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-sdk/src/blocks.ts:36]: Confirmed on all three points and fixed in 7663b1a11.

Verified against the file before changing anything: `STABILITY.md` opens with "This document is the authoritative ledger", states "Every export is annotated with a TSDoc release tag (`@public` / `@experimental` / `@deprecated`) that mirrors the table below", and closes the loop with "Everything not listed as `@public` is internal/unstable". Grepping it for `blocks` returned exactly one hit — inside the *Custom services* row's prose — so the subpath had no row of its own, and in `blocks.ts` only `defineBlock` carried a tag while the nine type exports carried none.

**What I did.** Added a *Block authoring (`@nextlyhq/plugin-sdk/blocks`)* row to the experimental table listing all nine exports, why it is a subpath, where the registration side lives instead, and the D55 graduation condition. Then tagged every one of the nine.

**One thing I found while fixing it that changes where the tags belong.** The tags cannot live in the SDK. `packages/plugin-sdk/dist/blocks.d.ts` is emitted by the declaration bundler as a single line:

```ts
export { AnyBlockDefinition, BlockDefinition, /* ... */ defineBlock } from '@nextlyhq/blocks-engine';
```

Every attached comment is dropped — which means the pre-existing `@experimental` on `defineBlock` was not reaching consumers either, and adding tags in `blocks.ts` alone would have looked like a fix while changing nothing an author sees. A consumer's editor resolves these through to `@nextlyhq/blocks-engine`, so that is where the tags now are, on the declarations themselves. Verified in the emitted artifact rather than the source: `packages/blocks-engine/dist/index.d.ts` now contains 9 of them, one per export.

Green: types, lint, sherif, 479 unit tests across the three packages, and the page-builder integration suite with a test-name set identical to the pre-change baseline.

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/plugin-page-builder/src/index.ts [hunk 17]
```diff
@@ -54,3 +54,26 @@ export type {
...
+export type { BlockRegistrationService } from "./blocks/registration-service";
+// `defineBlock` and `BlockDefinition` are NOT re-exported here. The package
+// root already exports both from `./core`, where `defineBlock` registers into
+// `defaultBlockRegistry` and `BlockDefinition` is the PoC's own shape. An
+// explicit re-export would shadow the star export, so a consumer importing
+// `defineBlock` from the root would silently get the engine's helper — which
+// only returns its argument — and stop registering its blocks.
+//
+// Block authors take `defineBlock` from `@nextlyhq/blocks-engine` directly
+// until the PoC registry is removed and the name is free.
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/index.ts]: Confirmed and fixed in edb960981.

The note did say exactly that — "Block authors take `defineBlock` from `@nextlyhq/blocks-engine` directly until the PoC registry is removed and the name is free". It predated the decision, taken later in this same PR, to publish the contracts at `@nextlyhq/plugin-sdk/blocks`, and it was never revisited. It now points at the SDK subpath, and notes that the collision it was working around does not exist there.

**You also surfaced a second stale statement in the same block, in the opposite direction.** The doc comment immediately above it read "`defineBlock` is re-exported here rather than reached through the core SDK" — describing a re-export that the very next comment explains is deliberately absent. That was the leftover of the earlier position I reversed when the SDK subpath was chosen, so the file was simultaneously claiming to re-export `defineBlock` and explaining why it must not. Both now say the same thing: registration lives here, the authoring contracts come from the SDK.

I swept for siblings before committing rather than fixing only the line you cited — no other guidance in `docs/`, the package READMEs, or the plugin sources tells an author to import from the engine.

Green: build, types, lint, 305 unit tests.

**Comment by mobeenabdullah** [comment]
@codex please review this PR
